Fix pickle handling for SingletonGate class

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit fixes an oversight in #10314 where the handling of pickle wasn't done correctly. Because the SingletonGate class defines __new__ and based on the parameters to the gate.__class__() call determines whether we get a new mutable copy or a shared singleton immutable instance we need special handling in pickle. By default pickle will call __new__() without any arguments and then rely on __setstate__ to update the state in the new object. This works fine if the original instance was a singleton but in the case of mutable copies this will create a singleton object instead of a mutable copy. To fix this a __reduce__ method is added to ensure arguments get passed to __new__ forcing a mutable object to be created in deserialization. Then a __setstate__ method is defined to correctly update the mutable object post creation.

### Details and comments

mtreinish committed:
Fix pickle handling for SingletonGate class

This commit fixes an oversight in #10314 where the handling of pickle
wasn't done correctly. Because the SingletonGate class defines __new__
and based on the parameters to the gate.__class__() call determines
whether we get a new mutable copy or a shared singleton immutable
instance we need special handling in pickle. By default pickle will call
__new__() without any arguments and then rely on __setstate__ to update
the state in the new object. This works fine if the original instance
was a singleton but in the case of mutable copies this will create a
singleton object instead of a mutable copy. To fix this a __reduce__
method is added to ensure arguments get passed to __new__ forcing a
mutable object to be created in deserialization. Then a __setstate__
method is defined to correctly update the mutable object post creation.

mtreinish committed:
Merge branch 'main' into fix-pickle-singletons

jakelishman commented on qiskit/circuit/singleton_gate.py:
```diff
@@ -80,6 +80,25 @@ def __init__(self, *args, _condition=None, **kwargs):
+            self.label = state[0]
+            self._condition = state[1]
+            self.duration = state[2]
+            self.unit = state[3]
```
If the `__setstate__` part of the `__reduce__` tuple is `None`, then `__setstate__` doesn't get called at all.  I think the check has no real meaning here.

wshanks replied:
The docs mention something like this for `__getstate__` [here](https://docs.python.org/3/library/pickle.html#object.__setstate__):

> Note: If [__getstate__()](https://docs.python.org/3/library/pickle.html#object.__getstate__) returns a false value, the [__setstate__()](https://docs.python.org/3/library/pickle.html#object.__setstate__) method will not be called upon unpickling.

but I didn't see it specifically called out for `__reduce__`. I confirmed that `__setstate__` was not called for `test_immutable_pickle`.

mtreinish replied:
This isn't relevant anymore with: https://github.com/Qiskit/qiskit/commit/3ad38b4ae95f23bae69d47942162345ed71ca6dd

jakelishman commented on qiskit/circuit/singleton_gate.py:
```diff
@@ -80,6 +80,25 @@ def __init__(self, *args, _condition=None, **kwargs):
+            self.__class__,
+            (self.label, self._condition, self.duration, self.unit),
+            (self.label, self._condition, self.duration, self.unit),
+        )
```
Did you have a particular reason to prefer `__setstate__`+`__reduce__` over `__getstate__`+`__setstate__`+`__getnewargs_ex__`?  The current form works fine so doesn't need changing, just asking because the docs suggest we should prefer the higher-level interface.

wshanks replied:
Actually, I wasn't that familiar with the details of pickling, so I read the docs closely and I think there is a problem with the current form.

When this 3-tuple form is used for reduce, the second element of the tuple is unpacked and passed as the args to the first tuple element. Then `__setstate__` is called on the object returned by that call with the third tuple element passed in. The issue is that it is not guaranteed that all subclasses of `SingletonGate` take `(label, _condition, duration, unit)` as their first four arguments. For example, `test_mutable_pickle` uses `SXGate` which takes `(label, duration, unit, _condition)` (note the order is not the same as `SingletonGate`). When `pickle.load()` is called in the test, the second element of the reduce tuple gets passed to `SXGate` and triggers the mutable creation in `__new__` but then also gets passed to `SXGate.__init__` and initially creates an object with `_condition` and `label` exchanged.

`__getnewargs_ex__` could work if it were guaranteed that all subclasses would accept `(label, _condition, duration, unit)` as keyword arguments. For something more flexible, maybe `__reduce__` could return something like `(object.__new__, (self.__class__,), (self.label, self._condition, self.duration, self.unit))` so that unpickling still bypasses `__init__` but creates a copy it can set the state on?

jakelishman replied:
Oh, that's true - I mistook that the first argument is the class object not specifically its `__new__`.  I think the reduction here would be fixed just by making the callable fetch `self.__class__.__new__` exactly to bypass `__init__`.  In practice, I don't think the current form will actually exhibit an issue, because the subsequent call to `__setstate__` will overwrite any mistakes during the `__init__`, but if the `__init__` started doing more type checking on its inputs, it could cause spurious exceptions before that stage.

I think the `__getstate__`/`__setstate__`/`__getnewargs_ex__` form might be a little easier - I think it's just the higher-level formalism of the system I described above with deliberately fetching `__class__.__new__`.

wshanks replied:
I played with `__reduce__` returning `(object.__new__, (self.__class__,), (self.label, self._condition, self.duration, self.unit))`. It kind of worked, but it helped me understand the point about pickle not calling `__init__`. `test_mutable_pickle` created a mutable `SXGate` and set the four attributes appropriately with `__setstate__`. However, the test then failed because the instance did not have a `_name` attribute because only those four attributes were set as `super().__init__()` was never called to set the other attributes. The current was passing the test because returning `self.__class__` from `__reduce__` leads to `__init__` getting called.

So now I understand why you advocate for the higher level methods. I think what would be perhaps most robust is defining `__getnewargs_ex__` to return a special keyword argument like `_create_mutable=True` (which hopefully no subclass would use) and then updating `__new__` to create a copy when this is passed (as well as the other conditions). Then we should not need `__getstate__` or `__setstate__` since the default implementation of copying attributes should be sufficient.

mtreinish replied:
I did it this way mostly because my knowledge of the pickle api is dated and I only knew how to do this with `__reduce__`. I wasn't aware of `__getnewargs_ex__`, but that interface is basically exactly what is needed here. I can refactor this to use the newer api.

mtreinish replied:
I updated this in: https://github.com/Qiskit/qiskit/pull/10871/commits/3ad38b4ae95f23bae69d47942162345ed71ca6dd we only need to define: `__getnewargs_ex__`  with that defined pickle will call `__new__` correctly so we either correctly use the singleton or the mutable copy returned. After that the default pickle mechanisms work fine as is.

jakelishman commented on test/python/circuit/test_singleton_gate.py:
```diff
@@ -251,3 +253,26 @@ def test_positional_label(self):
+    def test_immutable_pickle(self):
+        gate = SXGate()
+        self.assertFalse(gate.mutable)
+        with io.BytesIO() as fd:
+            pickle.dump(gate, fd)
+            fd.seek(0)
+            copied = pickle.load(fd)
```
Totally unimportant, but pickle allows serialisation to `bytes` directly, so you can do this as a one-liner if you prefer: `copied = pickle.loads(pickle.dumps(gate))`.

mtreinish replied:
I just don't like using `pickle.dumps` because `s` implies string to me and it's not a string :D

wshanks replied:
Rest in peace, Python 2 :slightly_smiling_face:

jakelishman replied:
maybe it's an initial from a language that reads right-to-left: "DUMP byteS" instead of "DUMP String" lol

wshanks replied:
Here's the Python 2 docstring for dumps :slightly_smiling_face: : "Return the pickled representation of the object as a string, instead of writing it to a file."

jakelishman commented on test/python/circuit/test_singleton_gate.py:
```diff
@@ -251,3 +253,26 @@ def test_positional_label(self):
         self.assertIsNot(gate, label_gate)
         self.assertEqual(label_gate.label, "I am a little label")
+
+    def test_immutable_pickle(self):
+        gate = SXGate()
+        self.assertFalse(gate.mutable)
+        with io.BytesIO() as fd:
+            pickle.dump(gate, fd)
+            fd.seek(0)
+            copied = pickle.load(fd)
+        self.assertFalse(copied.mutable)
```
I think with this test, we potentially might want to assert a step further and say `gate is copied`?  If we're pickling within the same process that should hold - obviously not so in IPC, but `is` comparisons and the memory reduction aren't relevant over that.

mtreinish replied:
Updated in: https://github.com/Qiskit/qiskit/pull/10871/commits/c24edc1867b68a0c08c0ea5be3561e359387e3d3

jakelishman reviewed (commented):
This looks sensible to me, thanks.  The only real comment is the last one about potentially strengthening what we want to assert about the within-process pickle roundtrip.

wshanks commented:
Here is some context for this issue. It was first noticed when the daily run of qiskit-experiments' tests with qiskit's main branch encountered test failures for circuits with conditioned gates (see for example [this run](https://github.com/Qiskit-Extensions/qiskit-experiments/actions/runs/6243811511/job/16949781007)). The tests only failed on Linux because on Windows and macOS the transpiler parallelism is disabled. When parallelism is used, objects are passed between processes using pickle, triggering this issue.

The issue is a little bit worse than mutable gates becoming instances of the immutable singleton when unpickled. The default unpickling procedure given in the documentation is:

```python
def restore(cls, attributes):
    obj = cls.__new__(cls)
    obj.__dict__.update(attributes)
    return obj
```

Every `cls.__new__(cls)` will generate the singleton gate instance, but note that the `obj.__dict__.update(attributes)` will actually mutate the immutable instance, so you could end up with an unexpected condition on all the default `SXGate`s for example.

jakelishman commented:
Defining `__reduce__` or `__setstate__` prevents that default `restore` from taking effect, I'm pretty sure - that's kind of the point of the methods.

wshanks commented on test/python/circuit/test_singleton_gate.py:
```diff
@@ -251,3 +253,26 @@ def test_positional_label(self):
+            copied = pickle.load(fd)
+        self.assertFalse(copied.mutable)
+
+    def test_mutable_pickle(self):
```
One thing that surprised me with this issue is that it was not caught in this repo before the singleton gate PR was merged. I expect there is a good set of tests for transpiling conditional gates, but likely the tests all transpile a single circuit and don't hit the pickling issues that come up with parallel transpilation. I wonder if the tests added here are enough or if there should be more tests of pickling and unpickling circuits.

jakelishman replied:
There's actually not a _vast_ amount of testing of conditional gates (with `.condition` set), because they typically weren't executable on anything but simulators until IBM hardware grew support for midcircuit measurements and more generalised control flow, at which point we began pivoting to `IfElseOp`.

The bug getting through points to a deficiency in integration testing around conditional gates for sure, and we can add more tests along those lines until `Instruction.condition` is completely removed in 1.0.  Conditionals after transpilation are also harder to write good tests about, because we can't just test for unitary equivalence or anything like that when they're present, so it's easier for stuff to slip through.

wshanks commented:
Right, I was just describing how things work now where there is no `__setstate__` or `__reduce__`.

mtreinish committed:
Use __getnewargs_ex__ insetad of __reduce__ & __setstate__

This commit pivots the pickle interface methods used to implement
__getnewargs_ex__ instead of the combination of __reduce__ and
__setstate__. Realistically, all we need to do here is pass that
we have mutable arguments to new to trigger it to create a separate
object, the rest of pickle was working correctly. This makes the
interface being used for pickle a lot clearer.

mtreinish committed:
Improve assertion in immutable pickle test

mtreinish committed:
Merge branch 'main' into fix-pickle-singletons

jakelishman reviewed (approved):
As far as I'm concerned this seems good to go, but I don't know if Will would like to push for more integration tests like he mentioned.

wshanks reviewed (approved):
I think this is okay as it is. I will open an issue about possible testing. I think the tests added here cover the current cases of concern. What I worry about is someone adding a new `SingletonGate` subclass with an extra slot/attribute not captured in `__getnewargs_ex__`. It might appear fine in single process tests but then misbehave in user code once it starts getting run through pickle for multiprocess transpiling.
